packages/expo-updates/ios/EXUpdates/AppLoader/EXUpdatesCodeSigningConfiguration.swift

```diff
@@ -0,0 +1,138 @@
+    }
+    
+    let extendedKeyUsage = x509Certificate.extendedKeyUsage
+    if (!extendedKeyUsage.contains(EXUpdatesCodeSigningConfiguration.EXUpdatesCodeSigningCertificateExtendedUsageCodeSigningOID)) {
+      throw EXUpdatesCodeSigningConfigurationError.CertificateMissingCodeSigningError
+    }
+    
+    keyId = metadata[EXUpdatesCodeSigningMetadataFields.KeyIdFieldKey] ?? EXUpdatesSignatureHeaderInfo.DefaultKeyId
+    algorithm = try parseCodeSigningAlgorithm(str: metadata[EXUpdatesCodeSigningMetadataFields.AlgorithmFieldKey])
+  }
+  
+  @objc
+  public func createAcceptSignatureHeader() -> String {
+    return "sig, keyid=\"\(keyId)\", alg=\"\(algorithm.rawValue)\""
+  }
+  
+  @objc
+  public func verifySignatureHeaderInfo(signatureHeaderInfo: EXUpdatesSignatureHeaderInfo, signedData: Data) throws -> NSNumber {
+    // check that the key used to sign the response is the same as the key embedded in the configuration
+    // TODO(wschurman): this may change for child certificates and development certificates
+    if (signatureHeaderInfo.keyId != self.keyId) {
+      throw EXUpdatesCodeSigningConfigurationError.KeyIdMismatchError
+    }
+    
+    // note that a mismatched algorithm doesn't fail early. it still tries to verify the signature with the
+    // algorithm specified in the configuration
+    if (signatureHeaderInfo.algorithm != self.algorithm) {
+      NSLog("Key with alg=\(signatureHeaderInfo.algorithm) from signature does not match client configuration algorithm, continuing")
+    }
+    
+    guard let certificate = SecCertificateCreateWithData(nil, certificateDataDer as CFData) else {
+      throw EXUpdatesCodeSigningConfigurationError.CertificateParseError
+    }
+    guard let publicKey = SecCertificateCopyPublicKey(certificate) else {
+      throw EXUpdatesCodeSigningConfigurationError.CertificateParseError
+    }
+    
+    guard let signatureData = Data(base64Encoded: signatureHeaderInfo.signature) else {
+      throw EXUpdatesCodeSigningConfigurationError.SignatureEncodingError
+    }
+    
+    let isValid = try self.verifyRSASHA256SignedData(signedData: signedData, signatureData: signatureData, publicKey: publicKey)
+    return isValid ? NSNumber(booleanLiteral: true) : NSNumber(booleanLiteral: false)
+  }
+  
+  private func sha256(data : Data) -> Data {
+      var digest = Data(count: Int(CC_SHA256_DIGEST_LENGTH))
+      data.withUnsafeBytes { bytes in
+        digest.withUnsafeMutableBytes { mutableBytes in
+          _ = CC_SHA256(bytes.baseAddress, CC_LONG(data.count), mutableBytes.bindMemory(to: UInt8.self).baseAddress)
+        }
+      }
+      return digest
+  }
+  
+  private func verifyRSASHA256SignedData(signedData: Data, signatureData: Data, publicKey: SecKey) throws -> Bool {
+    let hashBytes = self.sha256(data: signedData)
+    var error: Unmanaged<CFError>?
+    return SecKeyVerifySignature(publicKey, .rsaSignatureDigestPKCS1v15SHA256, hashBytes as CFData, signatureData as CFData, &error)
```

Comment:
I assume so because of the tests but does this properly throw?

Reply:
Update: it seems that when the signature is invalid this also throws an error instead of just returning false. This could be tricky.